### PR TITLE
Started adding JSDoc

### PR DESCRIPTION
**Pull request by shashkovdanil**
Even though I love TypeScript, I respect the decision made by the owners. I can admit that sometimes TypeScript causes a headache and you have to use `as/any/unknown`. It's also hard to read and understand what is happening in code like this.

```ts
export type GetKeysByValue<T extends Values> = {
  [K in keyof ItemMap]: ItemMap[K] extends { value: infer U } ? (U extends T ? K : never) : never;
}[keyof ItemMap];
```

However, it must be acknowledged that the code's readability has decreased. It will be harder for users to understand how to use turbo, and for other contributors to understand what functions and methods are doing. And instead of arguing, I prefer to find compromises. Therefore, I decided to combine the best of both worlds and started adding JSDoc. You can see what this looks like in this PR. I don't want to waste my time in vain, so I'm putting this up for public judgment. If the maintainers approve this PR, I will cover the remaining code with JSDoc.

### CODE REVIEW

File: src/core/session.js
Comment (seanpdoyle) [inline src/core/session.js:122]: This is either [EventSource](https://developer.mozilla.org/en-US/docs/Web/API/EventSource) | [WebSocket](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket)

File: src/core/session.js
Comment (seanpdoyle) [inline src/core/session.js:131]: This is either [EventSource](https://developer.mozilla.org/en-US/docs/Web/API/EventSource) | [WebSocket](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket)

**Comment by seanpdoyle** [comment]
Thank you for opening this. I think this is an improvement over what's currently available. 

I'm in favor of documentation of any kind, especially documentation that brings typed specificity.

I worry that the core team's disinterest in TypeScript might extend to a disinterest in typed code comments. Even the most correctly typed JSDoc comments risk becoming useless without the necessary tooling to enforce them.

Could this pull request also introduce tooling for CI and local development?

Does the core team have interest in incorporating that tooling into the project?

**Comment by WagnerMoreira** [comment]
I'm highly interested in making contributions here and adding JSDoc over the whole codebase if this move is accepted. 

@seanpdoyle 
> Could this pull request also introduce tooling for CI and local development?

you mean things like using `eslint-plugin-jsdoc` and checking for the presence and correctness of JSDoc comments?

**Comment by FroggyPanda** [comment]
I would also like to help out with reading types for this project.  However, I would like one of the main contributors, like @.dhh, to say if they're willing to add this PR so our time doesn't get wasted reading something that was removed.

File: src/core/session.js
Comment (DASPRiD) [inline src/core/session.js:105]: Would it make sense to define the options properties? This can be done via [`@typedef`](https://jsdoc.app/tags-typedef.html).

File: src/core/session.js
Comment (DASPRiD) [inline src/core/session.js:165]: This could be typed as `@param {('on'|'off')} mode`.

File: src/core/session.js
Comment (DASPRiD) [inline src/core/session.js:422]: `submitter` is optional, it should therefore be typed as `@param {HTMLElement} [submitter]`.

**Comment by DASPRiD** [comment]
This is a good idea. I did check out your branch and ran tsc to generate type definitions from this, which produces quite usable results for the files you touched:

```bash
npx -p typescript tsc src/**/*.js --declaration --allowJs --emitDeclarationOnly --outDir types
```

**Comment by shashkovdanil** [comment]
Hey guys! So I'm waiting for answer from maintainers, if they're interested in it, I'll add infrastructure like eslint and continue writing documentation

**Comment by WebReflection** [comment]
imho, "*this is the way*" (cit)

the benefits of JSDoc TS are many:

  * the `tsc` build step is not strictly necessary anymore, but it can always be used to `--check` from time to time via `npx` without having it as dependency
  * JSDoc itself encourage better documentation, not just types, and there are tools able to use that documentation to generate out of the box proper static pages anyone can read or crawl whenever it's needed
  * if focused only on the public API, the rest of the code will be still enjoyable (for the team) and maintenance would be kept minimal as public API signatures are all dependent projects need to have TS work, as [TS works already very well with JSDoc](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html)

last, but not least, I love the fact somebody proposed something instead of insulting people involved in this project (not everyone, but what I've read these days show how TS community can be way more toxic and hostile than anything happened in here).

I hope maintainers will give this PR a chance.

**Review by w3bdesign — APPROVED** [review]

**Comment by afcapel** [comment]
@shashkovdanil thanks for the civil and constructive approach to the PR, I really appreciate it.

What we found that works for us to document Turbo is having a [documentation project](https://github.com/hotwired/turbo-site) that you can [access online](https://turbo.hotwired.dev/reference/drive).

Most of the public Turbo APIs are HTML annotations (data- attributes, mostly), custom elements and events. Having a documentation website allows us to document all of those together in a consistent way.

There are a few cases where you might want to call a Turbo public API directly from your code, like `Turbo.visit` or `Turbo.cache.clear`. For those cases we were thinking to ship some type annotations to [DefinitelyTyped](https://github.com/DefinitelyTyped), so people that like TS can enjoy their types.

**Comment by shashkovdanil** [comment]
@afcapel Hi, JSDoc is not only good for users who will use the public API, but it is also good for contributors. If the documentation covers all this and JSDoc is not useful, I think you can close the PR

**Review by ArnavK-09 — APPROVED** [review]

**Review by FroggyPanda — APPROVED** [review]

**Comment by Gravy59** [comment]
Hi, you might consider using [`ts-to-jsdoc`](https://github.com/futurGH/ts-to-jsdoc) or similar to expedite the process.

**Comment by marcoroth** [comment]
Here's an attempt to migrate the whole codebase to JSDoc using the above mentioned `ts-to-jsdoc` package: https://github.com/marcoroth/turbo/commit/1ea3bfc3c061e693dffd9a597929479be254519d

The generated code is not 100% perfect in all cases, but I would be more than happy to polish this by hand if we decide to continue moving forward with this.

I think having the comments/documentation is valuable for contributors to see and understand how the internals work. This is also why it's not really a solution to just push up the types to DefinitelyTyped. It's not too valuable for users using Turbo and consuming the exported types the packages ships with, it's about the context and documentation we lost about the internals.

**Review by TechStudent10 — APPROVED** [review]
This is a great step forward. If the devs (looking at you DHH) really want to ditch TypeScript, I think JSDoc is a great in-between

**Comment by shashkovdanil** [comment]
@afcapel hey, should I continue this?

**Comment by ChristianECG** [comment]
image

**Comment by afcapel** [comment]
@shashkovdanil I'm afraid we're not interested in bringing back the type system, even if it's in the form of comments.

But thanks again for the constructive approach to the discussion.

**Pull request closed without merging**
